File: classes/ListVC.m
//
//  ListVC.m
//  mdxplayer
//
//  Created by asada on 2013/04/04.
//  Copyright (c) 2013 asada. All rights reserved.
//

#import "AppDelegate.h"
#import "ListVC.h"
#import "Player.h"
#import "CFileManager.h"
#import "InfoVC.h"

#import <Dropbox/Dropbox.h>




@interface ListVC ()

@end

@implementation ListVC
{
    NSMutableArray *files;
    NSMutableArray *folders;
    UIBarButtonItem *rbb;
    NSMutableDictionary *file2title;
}

-(void)dealloc
{
	[NCenter removeObserver:self];
}

- (void)viewDidLoad
{
    [super viewDidLoad];
	self.tableView.separatorStyle = UITableViewCellSeparatorStyleNone;
	self.tableView.backgroundColor = [UIColor lightGrayColor];
	
	UIBarButtonItem *bb = [[UIBarButtonItem alloc] init];
	bb.tintColor = [UIColor colorWithWhite:0.9 alpha:1];// colorWithWhite:0.8 alpha:1];
	[bb setTitleTextAttributes:@{UITextAttributeTextColor:[UIColor darkGrayColor],UITextAttributeTextShadowColor:[UIColor clearColor]} forState:UIControlStateNormal];
	self.navigationItem.backBarButtonItem = bb;
	
	[self.navigationItem.rightBarButtonItem setTitleTextAttributes:@{UITextAttributeFont:[UIFont boldSystemFontOfSize:17]} forState:UIControlStateNormal];

    rbb = self.navigationItem.rightBarButtonItem;
	
    if(!_path)	// root
    {
        self.title = @"MDX";
        UIButton *btn = [UIButton buttonWithType:UIButtonTypeInfoDark];
        [btn addTarget:self action:@selector(tapInfo:) forControlEvents:UIControlEventTouchUpInside];
        UIBarButtonItem *bbb = [[UIBarButtonItem alloc] initWithCustomView:btn];
        self.navigationItem.leftBarButtonItem = bbb;
        
    }

	if([_path isEqualToString:cfmFolderDropbox])
	{
		[self redispDropboxRB];
		[NCenter addObserver:self selector:@selector(reload) name:cfmNotifyReload object:nil];
		[NCenter addObserver:self selector:@selector(dropbox_logined) name:@"DROPBOX_LOGINED" object:nil];
	}
	
    file2title = [NSMutableDictionary dictionary];
    [self reload];
}

#pragma mark Dropbox assists

-(void)redispDropboxRB
{
    
    UIBarButtonItem *sbb;
    if([DBAccountManager sharedManager].linkedAccount)
        sbb = [[UIBarButtonItem alloc] initWithTitle:@"logout" style:UIBarButtonItemStyleBordered target:self action:@selector(dropbox_logout)];
    else
        sbb = [[UIBarButtonItem alloc] initWithTitle:@"login" style:UIBarButtonItemStyleBordered target:self action:@selector(dropbox_login)];

    sbb.tintColor = [UIColor colorWithWhite:0.9 alpha:1];
    [sbb setTitleTextAttributes:@{UITextAttributeTextColor:[UIColor darkGrayColor],UITextAttributeTextShadowColor:[UIColor clearColor]} forState:UIControlStateNormal];

    self.navigationItem.rightBarButtonItems = @[rbb,sbb];
}

-(void)dropbox_login
{
    [[DBAccountManager sharedManager] linkFromController:self];
}

-(void)dropbox_logout
{
    [[DBAccountManager sharedManager].linkedAccount unlink];
    folders = nil;
    files = nil;
    [self.tableView reloadData];
}

-(void)dropbox_logined
{
	[self redispDropboxRB];
	[self reload];
}

#pragma mark rutines

-(void)reload
{
    [CFileMnager listOfFolder:_path completion:^(NSArray *lists) {
		
		
		if(!lists || !lists.count)
		{
			if([_path isEqualToString:cfmFolderDropbox])
				[self showMessage:@"Copy files to Dropbox /Apps/mdx or /アプリ/mdx"];
			
			if([_path isEqualToString:cfmFolderDocuments])
				[self showMessage:@"Copy files with iTunes app from PC/mac/X68k"];
			
			return;
		}
		
        files = [NSMutableArray array];
        folders = [NSMutableArray array];

        for(CFileInfo *info in lists)
        {
            if(info.isFolder)
            {
                [folders addObject:info];
                continue;
            }
            
            if(![info.title.lowercaseString hasSuffix:@".mdx"]) continue;
            [files addObject:info];
        }
        [self.tableView reloadData];
    }];
    
}

-(PlayerVC*)getPlayerVC
{
    AppDelegate *gate = (AppDelegate*)[UIApplication sharedApplication].delegate;
    return gate.playerVC;
}

#pragma mark actions

-(void)tapInfo:(id)sender
{
	InfoVC *vc = [[InfoVC alloc] init];
    [self.navigationController pushViewController:vc animated:YES];
    
}

- (IBAction)tapPlayer:(id)sender
{
    PlayerVC *vc = [self getPlayerVC];
    [self.navigationController pushViewController:vc animated:YES];
    
}

-(void)showMessage:(NSString*)msg
{
	[[[UIAlertView alloc] initWithTitle:nil message:msg
							   delegate:nil cancelButtonTitle:@"OK" otherButtonTitles: nil] show];
}


#pragma mark table

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView
{
    return 2;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section
{
    if(!section) return folders.count;
    return files.count;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath
{
    static NSString *CellIdentifier = @"Cell";
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:CellIdentifier];
    if(!cell)
    {
        cell = [[UITableViewCell alloc] initWithStyle:UITableViewCellStyleSubtitle reuseIdentifier:CellIdentifier];
        cell.textLabel.font = [UIFont boldSystemFontOfSize:14];
        cell.textLabel.minimumScaleFactor = 0.6;
        cell.textLabel.numberOfLines = 2;
		
		
		cell.backgroundView = [[UIImageView alloc] initWithImage:[[UIImage imageNamed:@"tblbar"] resizableImageWithCapInsets:UIEdgeInsetsMake(4, 0, 4, 0) ]];
		cell.backgroundView.autoresizingMask = UIViewAutoresizingFlexibleHeight;
		
		cell.textLabel.backgroundColor = [UIColor clearColor];
		cell.detailTextLabel.backgroundColor = [UIColor clearColor];
		cell.textLabel.textColor = [UIColor colorWithWhite:0.25 alpha:1];
		cell.textLabel.shadowColor = [UIColor colorWithWhite:1 alpha:1];
		cell.textLabel.shadowOffset = CGSizeMake(.5, .5);
		
		cell.detailTextLabel.font = [UIFont systemFontOfSize:12];
		cell.selectionStyle = UITableViewCellSelectionStyleGray;
    }

    if(!indexPath.section)
    {
        CFileInfo *info = [folders objectAtIndex:indexPath.row];
        cell.textLabel.font = [UIFont boldSystemFontOfSize:17];
        cell.textLabel.text = info.title;
        cell.accessoryType = UITableViewCellAccessoryDisclosureIndicator;
        cell.detailTextLabel.text = nil;
    }
    else
    {
        CFileInfo *info = [files objectAtIndex:indexPath.row];
        cell.textLabel.font = [UIFont boldSystemFontOfSize:14];
        cell.detailTextLabel.text = info.title;
        cell.accessoryType = UITableViewCellAccessoryNone;

        NSString *title = [file2title objectForKey:info.path];
        if(title)
            cell.textLabel.text = title;
        else
        {
			// ここがあるからreuseダメかもね
            __weak UITableViewCell *wc = cell;
            [CFileMnager dataOfFile:info.path completion:^(NSData *data) {
                NSString *title = [Player titleForMDXData:data];
                wc.textLabel.text = title;
                [file2title setValue:title forKey:info.path];
                [wc setNeedsLayout];
            }];
        }
    }
    
    return cell;
}

-(CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath
{
    return indexPath.section ? 64 : 44;
}


- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath
{
    if(!indexPath.section)
    {
        ListVC *vc = [self.storyboard instantiateViewControllerWithIdentifier:@"list"];
        CFileInfo *info = [folders objectAtIndex:indexPath.row];
        vc.path = info.path;
        vc.title = info.title;
        [self.navigationController pushViewController:vc animated:YES];
        return;
    }
    
    PlayerVC *vc = [self getPlayerVC];
    
    CFileInfo *info = [files objectAtIndex:indexPath.row];
    [vc playFile:info.path path:_path];
    
    [self.navigationController pushViewController:vc animated:YES];
    
    
}

@end

File: classes/ListVC.swift
//
//  ListVC.swift
//  mdxplayer
//
//  Created by asada on 2016/08/13.
//  Copyright asada. All rights reserved.
//

import UIKit

class ListCell: UITableViewCell {

	required init?(coder _: NSCoder) {
		fatalError() }

	override init(style _: UITableViewCellStyle, reuseIdentifier: String?) {
		super.init(style: .subtitle, reuseIdentifier: reuseIdentifier)
		backgroundColor = UIColor(white: 23 / 255, alpha: 1)

		textLabel?.font = UIFont(name: "KH-Dot-Kodenmachou-16-Ki", size: 16) // UIFont.systemFontOfSize(16)
		textLabel?.font = .systemFont(ofSize: 16)
		textLabel?.textColor = .white
		textLabel?.numberOfLines = 2
		detailTextLabel?.textColor = UIColor(red: 174 / 255.0, green: 189 / 255.0, blue: 203 / 255.0, alpha: 1)
		detailTextLabel?.font = .systemFont(ofSize: 12)

		let sv = UIView()
		sv.backgroundColor = UIColor(white: 1, alpha: 0.2)
		selectedBackgroundView = sv
	}
}

struct Item {
	var title: String?
	var file = ""
	var isDir = false

	var json: NSObject {
		var r: [String: NSObject] = [:]
		r["file"] = file as NSObject
		r["isDir"] = isDir as NSObject
		r["title"] = (title ?? "") as NSObject
		return r as NSObject
	}

	init() {}

	init(file: String, isDir: Bool) {
		self.file = file
		self.isDir = isDir
		if isDir { title = file }
	}

	init(title: String, file: String, isDir: Bool) {
		self.title = title
		self.file = file
		self.isDir = isDir
	}

	init?(json: NSObject) {
		guard let r = json as? [String: NSObject] else { return nil }
		guard let f = r["file"] as? String else { return nil }
		file = f
		isDir = r["isDir"] as? Bool ?? false
		title = r["title"] as? String
		if title == "" { title = nil }
	}
}

class ListVC: UITableViewController {

	var path = ""
	var list: [Item] = []

	var localPath = ""

	static func path2local(_ p: String) -> String {

		let reps: [String: String] = [
			"_documents_": Path.documents,
			"_resources_": Path.resource("buildin"),
			"_dropbox_": Path.caches("_dropbox_"),
		]

		for (k, v) in reps {
			if p.hasPrefix(k) { return p.replace(k, v) }
		}
		return p
	}

	override func viewDidLoad() {
		super.viewDidLoad()
		localPath = ListVC.path2local(path)
		print(localPath)
		print(path)
		tableView.rowHeight = 66
		tableView.separatorColor = UIColor(white: 61 / 255, alpha: 1)
		tableView.backgroundColor = UIColor(white: 13 / 255, alpha: 1)
		tableView.register(ListCell.self, forCellReuseIdentifier: "cell")
		tableView.contentInset.bottom = 66
		reload()
	}

	func reload() {
		list.removeAll()
		let ar = Path.files(localPath)
		for s in ar {
			if s.hasPrefix(".") { continue }
			let isdir = Path.isDir(localPath.appendPath(s))
			if !isdir && !s.lowercased().hasSuffix(".mdx") { continue }
			list.append(Item(file: s, isDir: isdir))
		}
		sortedReload()

		if list.count == 0 && path == "_documents_" {
			showInfo("Copy files with iTunes app from PC/mac/X68k")
		}
	}

	func sortedReload() {
		list.sort { (a, b) -> Bool in
			if a.isDir != b.isDir { return a.isDir }
			return a.file.lowercased() < b.file.lowercased()
		}
		tableView.reloadData()
	}

	override func numberOfSections(in _: UITableView) -> Int {
		return 1
	}

	override func tableView(_: UITableView, numberOfRowsInSection _: Int) -> Int {
		return list.count
	}

	override func tableView(_: UITableView, heightForRowAt indexPath: IndexPath) -> CGFloat {
		if indexPath.row >= list.count { return 44 }
		let item = list[indexPath.row]
		return item.isDir ? 44 : 66
	}

	override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
		let cell = tableView.dequeueReusableCell(withIdentifier: "cell", for: indexPath)

		if indexPath.row >= list.count { return cell }

		let item = list[indexPath.row]

		if item.isDir {
			cell.textLabel?.text = item.title
			cell.detailTextLabel?.text = ""
			cell.accessoryType = .disclosureIndicator
			return cell
		}

		cell.textLabel?.text = item.file
		cell.accessoryType = .none

		cell.detailTextLabel?.text = item.file

		if let n = item.title {
			cell.textLabel?.text = n
			return cell
		}

		Dispatch.background {
			let n = Player.title(forMDXFile: self.localPath.appendPath(item.file))
			if indexPath.row >= self.list.count { return }
			self.list[indexPath.row].title = n
			Dispatch.main {
				guard let mc = tableView.cellForRow(at: indexPath) else { return }
				mc.textLabel?.text = n
			}
		}

		return cell
	}

	override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
		tableView.deselectRow(at: indexPath, animated: true)

		if indexPath.row >= list.count { return }
		let item = list[indexPath.row]

		if item.isDir {
			let vc = ListVC()
			vc.path = path.appendPath(item.file)
			vc.title = item.title
			navigationController?.pushViewController(vc, animated: true)
			return
		}

		doPlay(indexPath.row)
	}

	func doPlay(_ row: Int) {
		if row >= list.count { return }
		let item = list[row]
		var playlist: [String] = []
		var selected: Int = 0
		for c in list {
			if c.isDir { continue }
			if c.file == item.file { selected = playlist.count }
			playlist.append(localPath.appendPath(c.file))
		}

		HistoryListVC.addHistory(path, title: title ?? "")
		Player.sharedInstance().playFiles(playlist, index: selected)
	}
}
